projects/extension/sql/idempotent/001-openai.sql

```diff
@@ -19,196 +19,391 @@ set search_path to pg_catalog, pg_temp
+        'default_query': ai.openai.process_json_input(_default_query),
+        'http_client': ai.openai.process_json_input(_http_client),
+        '_strict_response_validation': _strict_response_validation
+    })
+
+    if 'config' not in GD['openai_client'] or ai.openai.client_config_changed(GD['openai_client']['config'], new_config):
+        client = ai.openai.make_async_client(plpy, **new_config)
+        GD['openai_client'] = {
+            'client': client,
+            'config': new_config
+        }
+
+$python$ LANGUAGE plpython3u VOLATILE SECURITY DEFINER PARALLEL UNSAFE;
+
+-------------------------------------------------------------------------------
+-- openai_client_destroy
+-- remove the client object stored in the global dictionary for the session
+CREATE OR REPLACE FUNCTION ai.openai_client_destroy() RETURNS void AS $python$
+    #ADD-PYTHON-LIB-DIR
+    if 'openai_client' in GD:
+        del GD['openai_client']
+$python$ LANGUAGE plpython3u VOLATILE SECURITY DEFINER PARALLEL UNSAFE;
+
 -------------------------------------------------------------------------------
 -- openai_list_models
 -- list models supported on the openai platform
 -- https://platform.openai.com/docs/api-reference/models/list
-create or replace function ai.openai_list_models(api_key text default null, api_key_name text default null, base_url text default null)
-returns table
-( id text
-, created timestamptz
-, owned_by text
-)
+create or replace function ai.openai_list_models(
+    _api_key text DEFAULT NULL,
+    _api_key_name text DEFAULT NULL,
+    _base_url text DEFAULT NULL,
+    _extra_headers jsonb DEFAULT NULL,
+    _extra_query jsonb DEFAULT NULL,
+    _extra_body jsonb DEFAULT NULL,
+    _timeout float8 DEFAULT NULL
+) returns jsonb
 as $python$
     #ADD-PYTHON-LIB-DIR
     import ai.openai
-    for tup in ai.openai.list_models(plpy, api_key, api_key_name, base_url):
-        yield tup
+    import json
+
+    # Create async client
+    client = ai.openai.get_or_create_client(plpy, GD, _api_key, _api_key_name, _base_url)
+
+    # Prepare kwargs for the API call
+    kwargs = {}
+    # Add extra parameters if provided
+    if _extra_headers is not None:
+        kwargs['extra_headers'] = json.loads(_extra_headers)
+    if _extra_query is not None:
+        kwargs['extra_query'] = json.loads(_extra_query)
+    if _extra_body is not None:
+        kwargs['extra_body'] = json.loads(_extra_body)
+
+    async def async_openai_call(client, kwargs):
+        response = await client.models.with_raw_response.list(**kwargs)
+        return response.text
+
+    # Execute the API call with cancellation support
+    result = ai.openai.execute_with_cancellation(plpy, client, async_openai_call, **kwargs)
+
+    return result
 $python$
-language plpython3u volatile parallel safe security invoker
-set search_path to pg_catalog, pg_temp
+    language plpython3u volatile parallel unsafe security invoker
```

Comment:
why are these functions parallel unsafe? My understanding is every parallel worker would get a different process and thus a different python interpreter. I don't think even GD is shared.

Reply:
Sorry, another artifact from my prior implementation. I was thinking it was going to be unsafe even if each call created it's own client due to overloading the API unintentionally, but in retrospect, it's better to control that with the standard GUC for controlling parallelism.